engine/itest/src/test/java/org/opennms/oce/engine/itest/topology/UpdateInventoryTest.java

```diff
@@ -0,0 +1,151 @@
+        topologyEngineFactory = new TopologyEngineFactory();
+
+        ModelBuilderImpl modelBuilder = new ModelBuilderImpl();
+        model = modelBuilder.buildModel();
+    }
+
+    /**
+     *      <!-- Devices -->
+     *      <model-object-entry type="Device" id="n3" parent-type="Model" parent-id="model"/>
+     *
+     *      <!-- Cards -->
+     *      <model-object-entry type="Card" id="n3-c1" parent-type="Device" parent-id="n3" />
+     *
+     *      <!-- Ports -->
+     *      <model-object-entry type="Port" id="n3-c1-p1" parent-type="Card" parent-id="n3-c1" />
+     *      <model-object-entry type="Port" id="n3-c1-p2" parent-type="Card" parent-id="n3-c1" />
+     *      <model-object-entry type="Port" id="n3-c1-p3" parent-type="Card" parent-id="n3-c1" />
+     *      <model-object-entry type="Port" id="n3-c1-p4" parent-type="Card" parent-id="n3-c1" />
+    */
+    @Test
+    public void canAddModelObject() {
+        /** Case: Add new Network Element (simple case adding new device with one card and four ports)
+         *
+         * Expected Results:
+         * - A new device is added with all its children (card and ports)
+         * - All relationships are established such as:
+         * -- A new device has parent
+         * -- A new device has peers (if applicable)
+         */
+
+        //Assert that inventory and model are constructed properly
+        ModelObject root = model.getRoot();
+        assertThat(root, notNullValue());
+        assertThat(root.getParent(), nullValue());
+        int topLevelOfToplogyModelCount = root.getChildren().size();
+
+        //and have all levels of model object hierarchy (device, card, port)
+        assertThat(model.getTypes(), hasItem("Device"));
+        assertThat(model.getTypes(), hasItem("Card"));
+        assertThat(model.getTypes(), hasItem("Port"));
+
+        //Construct new device with one card and four ports
+        final ModelObjectImpl device = new ModelObjectImpl("Device", "n3");
+
+        final ModelObjectImpl card = new ModelObjectImpl("Card", "n3-c1");
+
+        final ModelObjectImpl port1 = new ModelObjectImpl("Port", "n3-c1-p1");
+        final ModelObjectImpl port2 = new ModelObjectImpl("Port", "n3-c1-p2");
+        final ModelObjectImpl port3 = new ModelObjectImpl("Port", "n3-c1-p3");
+        final ModelObjectImpl port4 = new ModelObjectImpl("Port", "n3-c1-p4");
+
+        card.setParent(device);
+
+        port1.setParent(card);
+        port2.setParent(card);
+        port3.setParent(card);
+        port4.setParent(card);
+
+        /** Currently I place updating functionality in ModelImpl where model belongs to but:
+         * Design question: should updating model flow be delegated to ModelBuilder which is responsible for constructing initial model
+         * It can be passed as reference to another build function (updateModel) function which is similar to existing buildModel
+         */
+
+        ((ModelImpl)model).addObject(device);
```

Comment:
We can update the `ModelBuilderImpl` to return a `ModelImpl` instead of having to cast this every time..

Reply:
yes, Test work still in progress, the focus of this PR is to provide API.